pepper-apis/src/main/java/org/broadinstitute/ddp/model/copy/CompositeCopyConfigurationPair.java

```diff
@@ -0,0 +1,59 @@
+package org.broadinstitute.ddp.model.copy;
+
+import org.jdbi.v3.core.mapper.reflect.ColumnName;
+import org.jdbi.v3.core.mapper.reflect.JdbiConstructor;
+
+public class CompositeCopyConfigurationPair {
+
+    private long id;
+    private long copyConfigPairId;
+    private long sourceChildQuestionStableCodeId;
+    private String sourceChildQuestionStableId;
+    private long targetChildQuestionStableCodeId;
+    private String targetChildQuestionStableId;
+
```

Comment:
One reason I did this was that it makes the queries simpler. If I went with `CopyAnswerLocation` for child source/target, I will end up with like 4 more joins in the query just to get the question_stable_ids... I think I'll opt to keep it as a String.

I'm not sure I understand your suggested consideration, can you elaborate?

Reply:
Yes there's some duplicate notions. I think this helps a bit in containing the complexity of composites. Otherwise, I'm not quite sure how to handle composites.

Regarding your suggested consideration, I can't quite visualize how that will work. Can you elaborate?

Reply:
Incomplete (and you don't need to subclass if you don't want ti) , but I think you'll get the idea:
```java
class CompositeAnswerLocation extends CopyAnswerLocation {
    private long childQuestionStableCodeId;
    private String childQuestionStableId;
}
public class CopyConfigurationPair {

    private long id;
    private CopyLocation source;
    private CopyLocation target;
// can go away now
 //   private List<CompositeCopyConfigurationPair> compositeChildLocations = new ArrayList<>();
    private int order;
}
```
```xml
<createTable tableName="copy_answer_location">
            <column name="copy_location_id" type="bigint">
                <constraints primaryKey="true"
                             primaryKeyName="copy_question_location_pk"
                             references="copy_location(copy_location_id)"
                             foreignKeyName="copy_answer_location_copy_location_fk"
                             deleteCascade="true"/>
            </column>
            <column name="question_stable_code_id" type="bigint">
                <constraints nullable="false"
                             references="question_stable_code(question_stable_code_id)"
                             foreignKeyName="copy_answer_location_qsid_fk"/>
            </column>
           <!-- New column. Used only for composite -->
<column name="child_question_stable_code_id" type="bigint">
                <constraints nullable="true"
                             references="question_stable_code(question_stable_code_id)"
                             foreignKeyName="copy_answer_location_cqsid_fk"/>
            </column>
        </createTable>
```
Goes away:
```java
public class CompositeCopyConfigurationPair {
```
From CopyConfigurationDao
```java
private CopyLocation reduceLocation(RowView view, String prefix) {
            long locId = view.getColumn(columnName(prefix, "copy_location_id"), Long.class);
            var type = CopyLocationType.valueOf(view.getColumn(columnName(prefix, "copy_location_type"), String.class));
           Long childId = view.getColumn(columnName(prefix, "child_question_stable_code_id"), Long.class);
           // Not a composite!
            if (type == CopyLocationType.ANSWER && childId == null) {
                return new CopyAnswerLocation(locId,
                        view.getColumn(columnName(prefix, "question_stable_code_id"), Long.class),
                        view.getColumn(columnName(prefix, "question_stable_id"), String.class));
          // A composite!
           } else if (type 
            } else (type == CopyLocationType.ANSWER && childId != null) {
                     return new CompositeAnswerLocation(locId,
                        view.getColumn(columnName(prefix, "question_stable_code_id"), Long.class),
                        view.getColumn(columnName(prefix, "question_stable_id"), String.class),
                        childId);         
           }
                return new CopyLocation(locId, type);
            }
        }
```
etc.,etc.etc.

Reply:
Okay, I think I get the gist of it. This was something I had considered briefly before, but dismissed because I thought the complexity will be really high -- one drawback was that instead of having all the child in one contained list under the composite pairing, it's now "flattened" into the `CopyConfigurationPair` list so I thought it might be difficult to copy the composite answers. After fiddling with implementing the actual copying code, I think this might be do-able. I'll fiddle with the code a bit more before committing to this newer design.

Reply:
This suggestion is looking more appealing. Unless there's other opinions, I'm going to move to it.

Reply:
One additional thought: given that the way we create our study configurations, a child question can only belong to one parent composite question, we could possibly even dispense of creating an additional column for the child question id; provided with the child id we can figure out what is the parent id. This could also be carried forward to the .conf file; if you have composite child question that you want to use as a source or target no need to bother specifying the parent question id. The child question id is sufficient.
Just go with whatever makes more sense to you.

Reply:
Pushed up changes for new design.